{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/rdc143/GazelleProject/blob/main/InferFreq.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "Eh_Dhlkt3ts8"
      },
      "outputs": [],
      "source": [
        "!pip install pandas-plink\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "gAy6VuiU3YEY"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "from pandas_plink import read_plink\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "import multiprocessing\n",
        "import copy\n",
        "from scipy.optimize import fixed_point\n",
        "from scipy.stats import chi2\n",
        "from scipy.stats import uniform"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "NaLDCrVH6tK7"
      },
      "outputs": [],
      "source": [
        "from google.colab import drive\n",
        "drive.mount(\"/content/drive/\")\n",
        "\n",
        "%cd /content/drive/My Drive/Colab Notebooks/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "PSmjH1uukaKv"
      },
      "outputs": [],
      "source": [
        "### Load data\n",
        "(bim, fam, bed) = read_plink(\"./Gazelle/*ch16.bed\")\n",
        "\n",
        "fam[\"gender\"] = fam[\"gender\"].astype(int)\n",
        "sex = np.array(fam[\"gender\"]) / 2 # admixture proportions, proportion of autosomal linked(in contrast to y-linked)\n",
        "male = sex < 1\n",
        "\n",
        "missing = np.mean(np.isnan(bed), axis = 1)\n",
        "#sns.kdeplot(missing)\n",
        "#plt.show()\n",
        "\n",
        "freq = np.nanmean(bed, axis = 1) / 2\n",
        "#sns.kdeplot(freq)\n",
        "#plt.show()\n",
        "\n",
        "keep = (0.1 < freq) & (freq < 0.9) & (missing == 0) ## is first statement neccessary?\n",
        "#sns.kdeplot(keep)\n",
        "#plt.show()\n",
        "\n",
        "#np.random.seed(seed=42)\n",
        "\n",
        "geno = bed.compute()[keep] # allele counts for first allele, variant per row, individual per column\n",
        "#geno = 2 - geno  # flip all alleles\n",
        "for i in range(len(geno)): # flip alleles at random\n",
        "  if np.random.uniform() > 0.5:\n",
        "    geno[i] = 2 - geno[i]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "o6PLubbFKJ8b"
      },
      "outputs": [],
      "source": [
        "### Funcs\n",
        "def EMInferFreqs(x, geno, male, fixY=False, fixA=False):\n",
        "  f = copy.copy(x) # f[0] is f_Y and f[1] is f_A\n",
        "  G = np.array([np.where(geno > 1, 1, 0),\n",
        "                np.where(geno > 0, 1, 0)])[:,male] # ordered genotypes for males, TODO: replace with boolean, (do only for male?), do outside func?\n",
        "\n",
        "  p_YA = np.where(G[0] == 1, f[0], (1 - f[0])) * np.where(G[1] == 1, f[1], (1 - f[1]))\n",
        "\n",
        "  p_AY = np.where(G[0] == 1, f[1], (1 - f[1])) * np.where(G[1] == 1, f[0], (1 - f[0]))\n",
        "\n",
        "  q_YA = p_YA / (p_AY + p_YA)\n",
        "  q_AY = p_AY / (p_AY + p_YA)\n",
        "\n",
        "  E_c_Y = np.sum(q_AY * G[1]) + np.sum(q_YA * G[0])\n",
        "  E_C_Y = np.sum(q_AY * (1 - G[1])) + np.sum(q_YA * (1 - G[0]))\n",
        "\n",
        "  E_c_A = np.sum(geno[~male]) + np.sum(q_AY * G[0]) + np.sum(q_YA * G[1])\n",
        "  E_C_A = np.sum(2 - geno[~male]) + np.sum(q_AY * (1 - G[0])) + np.sum(q_YA * (1 - G[1]))\n",
        "\n",
        "  if fixY: return np.array([fixY,\n",
        "                   E_c_A / (E_c_A + E_C_A)])\n",
        "  if fixA: return np.array([E_c_Y / (E_c_Y + E_C_Y),\n",
        "                   fixA])\n",
        "\n",
        "  return np.array([E_c_Y / (E_c_Y + E_C_Y),\n",
        "                   E_c_A / (E_c_A + E_C_A)])\n",
        "\n",
        "\n",
        "def ll(f, G, male):\n",
        "  #to avoid problems with log computation and boundary of param space\n",
        "  if f[0] <= 1e-4: f[0] = 1e-4\n",
        "  if f[1] <= 1e-4: f[1] = 1e-4\n",
        "  if f[0] >= 1 - 1e-4: f[0] = 1 - 1e-4\n",
        "  if f[1] >= 1 - 1e-4: f[1] = 1 - 1e-4\n",
        "\n",
        "  isG = np.array([G == 0, G == 1, G == 2])\n",
        "  M = (np.sum(isG[0] & male), # counts of male\n",
        "       np.sum(isG[1] & male),\n",
        "       np.sum(isG[2] & male))\n",
        "  F = (np.sum(isG[0] & ~male), # counts of female\n",
        "       np.sum(isG[1] & ~male),\n",
        "       np.sum(isG[2] & ~male))\n",
        "\n",
        "  return (np.log((1 - f[0])*(1 - f[1])) * M[0] +\n",
        "          np.log((f[1]*(1 - f[0]) + f[0]*(1 - f[1]))) * M[1] +\n",
        "          np.log((f[0]*f[1])) * M[2] +\n",
        "          np.log(((1 - f[1])**2)) * F[0] +\n",
        "          np.log((2*(1 - f[1])*f[1])) * F[1] +\n",
        "          np.log((f[1]**2)) * F[2])\n",
        "\n",
        "\n",
        "def converge(geno, male, fixY=False, fixA=False, randomInit=False):\n",
        "  if np.any(np.isnan(geno)): raise Exception(\"Nan in genotypes >:(\") # check to make sure no nans are passed, even though at this point they should be filtered out\n",
        "  if randomInit:  # initialization\n",
        "    f = np.array([np.random.uniform(.1, .9), np.random.uniform(.1, .9)])\n",
        "  else:\n",
        "    f = np.array([0.5, 0.5])\n",
        "\n",
        "  change = np.array([float(\"Inf\"), float(\"Inf\")])\n",
        "  it = 0\n",
        "  while any(change > 1e-4) and it <= 500: # tolerance\n",
        "    fnew = EMInferFreqs(f, geno, male, fixY=fixY, fixA=fixA)\n",
        "    change = np.abs(f - fnew)\n",
        "    f = fnew\n",
        "    it +=1\n",
        "    if it > 500: print(f\"Max iterations reached {i}\")\n",
        "    #print((it, f))\n",
        "    #print((it, ll(f, geno, male)))\n",
        "  return f\n",
        "\n",
        "\n",
        "def sim(nsites=10_000, nind=10, frac_male=0.5, model=0, fixfrq=False):\n",
        "  nmale = int(np.rint(nind * frac_male))\n",
        "  male = np.array([True] * nmale + [False] * (nind - nmale))\n",
        "\n",
        "  geno = []\n",
        "  true_freqs=[]\n",
        "  for i in range(nsites):\n",
        "    f = [np.random.uniform(), np.random.uniform()] # sample f_Y and f_A\n",
        "    if model == 1:\n",
        "      f[1] = f[0]\n",
        "    if model == 2:\n",
        "      f[0] = np.random.choice([1e-4, 1 - 1e-4])\n",
        "    if model == 3:\n",
        "      f[1] = np.random.choice([1e-4, 1 - 1e-4])\n",
        "    if fixfrq:\n",
        "      f = fixfrq\n",
        "    true_freqs.append(f)\n",
        "\n",
        "    geno.append(np.random.binomial(1, f[1], nind) +\n",
        "                np.random.binomial(1, f[1], nind) * ~male +\n",
        "                np.random.binomial(1, f[0], nind) * male)\n",
        "\n",
        "  return np.array(geno), np.array(true_freqs), male\n",
        "\n",
        "\n",
        "def accelerate(geno, male): #not in use\n",
        "  if np.any(np.isnan(geno)): raise Exception(\"Nan in genotypes >:(\") # check to make sure no nans are passed, even though at this point they should be filtered out\n",
        "  init = np.array([0.5, 0.5])\n",
        "  args = (geno, male)\n",
        "  return fixed_point(EMInferFreqs, init, args=args, method=\"del2\", xtol=1e-4, maxiter=100)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "E_gD0W7n2LN4"
      },
      "outputs": [],
      "source": [
        "### Infer freqs and run tests\n",
        "freqs, freqs_null1, freqs_null2_0, freqs_null2_1, freqs_null3_0, freqs_null3_1 = [], [], [], [], [], []\n",
        "\n",
        "llik, llik_null1, llik_null2, llik_null3 = [], [], [], []\n",
        "\n",
        "for i in range(len(geno)):\n",
        "  freqs.append(converge(geno[i], male))\n",
        "  freqs_null1.append(np.mean(geno[i]) / 2)\n",
        "  freqs_null2_0.append(converge(geno[i], male, fixY=1e-4))\n",
        "  freqs_null2_1.append(converge(geno[i], male, fixY=1-1e-4))\n",
        "  freqs_null3_0.append(converge(geno[i], male, fixA=1e-4))\n",
        "  freqs_null3_1.append(converge(geno[i], male, fixA=1-1e-4))\n",
        "\n",
        "\n",
        "  llik.append(ll(freqs[i], geno[i], male))\n",
        "  llik_null1.append(ll(np.array([freqs_null1[i], freqs_null1[i]]), geno[i], male))\n",
        "  if ll(freqs_null2_0[i], geno[i], male) > ll(freqs_null2_1[i], geno[i], male):\n",
        "    llik_null2.append(ll(freqs_null2_0[i], geno[i], male))\n",
        "  else:\n",
        "    llik_null2.append(ll(freqs_null2_1[i], geno[i], male))\n",
        "\n",
        "  if ll(freqs_null3_0[i], geno[i], male) > ll(freqs_null3_1[i], geno[i], male):\n",
        "    llik_null3.append(ll(freqs_null3_0[i], geno[i], male))\n",
        "  else:\n",
        "    llik_null3.append(ll(freqs_null3_1[i], geno[i], male))\n",
        "\n",
        "freqs = np.array(freqs)\n",
        "freqs_null2 = np.array(freqs_null2_0) # dont really use these, remember to update to new null model if used\n",
        "freqs_null3 = np.array(freqs_null3_0)\n",
        "\n",
        "res = pd.DataFrame(np.array([np.array(bim[\"pos\"])[keep], freqs[:,0], freqs[:,1], freqs_null1, freqs_null2[:,1], freqs_null3[:,0], llik, llik_null1, llik_null2, llik_null3]).T,\n",
        "                   columns=[\"pos\", \"f_Y\", \"f_A\", \"f_null1\", \"f_null2\", \"f_null3\", \"log_likelihood\", \"log_likelihood_null1\", \"log_likelihood_null2\", \"log_likelihood_null3\"])\n",
        "\n",
        "res[\"test1\"] = 2 * (res[\"log_likelihood\"] - res[\"log_likelihood_null1\"])\n",
        "res[\"test2\"] = 2 * (res[\"log_likelihood\"] - res[\"log_likelihood_null2\"])\n",
        "res[\"test3\"] = 2 * (res[\"log_likelihood\"] - res[\"log_likelihood_null3\"])\n",
        "\n",
        "res[\"p1\"] = 1 - chi2.cdf(res[\"test1\"], df=1) + 1e-18 # added last term to avoid plotting difficulties on log scale\n",
        "res[\"p2\"] = 1 - chi2.cdf(res[\"test2\"], df=1) + 1e-18\n",
        "res[\"p3\"] = 1 - chi2.cdf(res[\"test3\"], df=1) + 1e-18\n",
        "\n",
        "res[\"logp1\"] = np.log10(res['p1'])\n",
        "res[\"logp2\"] = np.log10(res['p2'])\n",
        "res[\"logp3\"] = np.log10(res['p3'])\n",
        "\n",
        "#res[\"poly_shared\"] = ((0.03 < res[\"f_A\"]) & (res[\"f_A\"] < 0.97)) & ((0.03 < res[\"f_Y\"]) & (res[\"f_Y\"] < 0.97))\n",
        "\n",
        "alpha = 0.05\n",
        "res[\"sig2&3\"] = ( res[\"p2\"] < alpha ) & ( res[\"p3\"] < alpha )\n",
        "\n",
        "cutoff_cor = alpha / len(geno) # Bonferonni correction\n",
        "res[\"sig2&3_cor\"] = ( res[\"p2\"] < cutoff_cor ) & ( res[\"p3\"] < cutoff_cor )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "5G_sNyCDwcGI"
      },
      "outputs": [],
      "source": [
        "### plotting\n",
        "\n",
        "# distribution of frequencies\n",
        "sns.jointplot(res, x=\"f_Y\", y=\"f_A\", alpha = 0.05, marginal_ticks=True, ratio=2)\n",
        "plt.title(\"Joint distribution of inferred frequencies for chr 16\")\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res[res[\"pos\"] < 4.5e7], x=\"f_Y\", y=\"f_A\", alpha = 0.05, marginal_ticks=True, ratio=2)\n",
        "plt.title(\"Joint distribution of inferred frequencies for chr 16 in linked region\")\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res[res[\"pos\"] > 5.5e7], x=\"f_Y\", y=\"f_A\", alpha = 0.05, marginal_ticks=True, ratio=2)\n",
        "plt.title(\"Joint distribution of inferred frequencies for chr 16 in unlinked region\")\n",
        "plt.show()\n",
        "\n",
        "### test statistics\n",
        "#sns.jointplot(res, x=\"pos\", y=\"test1\", alpha=0.02)\n",
        "#plt.show()\n",
        "\n",
        "#sns.jointplot(res[res[\"pos\"] > 5.1e7], x=\"pos\", y=\"test1\", alpha=0.02)\n",
        "#plt.show()\n",
        "\n",
        "#sns.jointplot(res, x=\"pos\", y=\"test2\", alpha=0.02)\n",
        "#plt.show()\n",
        "\n",
        "#sns.jointplot(res[res[\"pos\"] > 5.1e7], x=\"pos\", y=\"test2\", alpha=0.02)\n",
        "#plt.show()\n",
        "\n",
        "#sns.jointplot(res, x=\"pos\", y=\"test3\", alpha=0.02)\n",
        "#plt.show()\n",
        "\n",
        "#sns.jointplot(res[res[\"pos\"] > 5.1e7], x=\"pos\", y=\"test3\", alpha=0.02)\n",
        "#plt.show()\n",
        "\n",
        "### p-values\n",
        "sns.jointplot(res[res[\"pos\"] > 5.1e7], x=\"pos\", y=\"logp1\")\n",
        "plt.axhline(np.log10(cutoff_cor), color=\"y\", lw=1.2)\n",
        "plt.title(\"p-values for LRT on null model 1 in unlinked region\", y=-0.16)\n",
        "plt.xlabel(\"position\")\n",
        "plt.ylabel(\"log10(p-value)\")\n",
        "plt.show()\n",
        "\n",
        "\n",
        "sns.jointplot(res, x=\"pos\", y=\"logp1\", alpha=0.02)\n",
        "plt.axhline(np.log10(cutoff_cor), color=\"y\", lw=1.2)\n",
        "plt.title(\"p-values for LRT on null model 1 (f_Y = f_A)\", y=-0.16)\n",
        "plt.xlabel(\"position\")\n",
        "plt.ylabel(\"log10(p-value)\")\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res, x=\"pos\", y=\"logp2\", alpha=0.02)\n",
        "plt.axhline(np.log10(cutoff_cor), color=\"y\", lw=1.2)\n",
        "plt.title(\"p-values for LRT on null model 2 (f_Y fixed)\", y=-0.16)\n",
        "plt.xlabel(\"position\")\n",
        "plt.ylabel(\"log10(p-value)\")\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res, x=\"pos\", y=\"logp3\", alpha=0.02)\n",
        "plt.axhline(np.log10(cutoff_cor), color=\"y\", lw=1.2)\n",
        "plt.title(\"p-values for LRT on null model 3 (f_A fixed)\", y=-0.16)\n",
        "plt.xlabel(\"position\")\n",
        "plt.ylabel(\"log10(p-value)\")\n",
        "plt.show()\n",
        "\n",
        "\n",
        "\n",
        "### sig 2 & 3\n",
        "sns.jointplot(res, x=\"f_Y\", y=\"f_A\", hue=\"sig2&3\", alpha = 0.2, marginal_ticks=True, ratio=2)\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res[res[\"pos\"] < 5.1e7], x=\"f_Y\", y=\"f_A\", hue=\"sig2&3\", alpha = 0.2, marginal_ticks=True, ratio=2)\n",
        "plt.title(\"Joint distribution of inferred frequencies for chr 16 in linked region\")\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res[(res[\"pos\"] < 5.1e7) & (res['f_Y'] <.99) & (res['f_Y'] >.01) & (res['f_A'] <.99) & (res['f_A'] >.01)], x=\"f_Y\", y=\"f_A\", hue=\"sig2&3\", alpha = 0.2, marginal_ticks=True, ratio=2)\n",
        "plt.title(\"Joint distribution of inferred frequencies for chr 16 in linked region\")\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res[res[\"pos\"] < 5.1e7], x=\"f_Y\", y=\"f_A\", hue=\"sig2&3_cor\", alpha = 0.2, marginal_ticks=True, ratio=2)\n",
        "plt.title(\"Joint distribution of inferred frequencies for chr 16 in linked region\")\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res[(res[\"pos\"] < 5.1e7) & (res['f_Y'] <.99) & (res['f_Y'] >.01) & (res['f_A'] <.99) & (res['f_A'] >.01)], x=\"f_Y\", y=\"f_A\", hue=\"sig2&3_cor\", alpha = 0.2, marginal_ticks=True, ratio=2)\n",
        "plt.title(\"Joint distribution of inferred frequencies for chr 16 in linked region\")\n",
        "plt.show()\n",
        "\n",
        "sns.histplot(res[res[\"sig2&3\"] & (res[\"pos\"] < 5.1e7)], x=\"pos\", bins=50)\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "-GHkrSUCbiEd"
      },
      "outputs": [],
      "source": [
        "### sim\n",
        "for simfrq in [[.6, .4], [.0, .7], [.7, .0], [.5, .5]]: # for main\n",
        "#for simfrq in [[.9, .05], [.05, .9], [.2, .2], [.9, .9], [1, .95], [.05, .05]]: # for appendix\n",
        "  simres = pd.DataFrame(columns=[\"f_Y\", \"f_A\", \"nind\"])\n",
        "  for n in [10, 25, 50, 100, 200]:\n",
        "\n",
        "    simgeno, true_freqs, simmale = sim(nind=n, fixfrq=simfrq)\n",
        "    simfreqs=[]\n",
        "    for i in range(len(simgeno)):\n",
        "      simfreqs.append(converge(simgeno[i], simmale))\n",
        "\n",
        "    inferred = pd.DataFrame(np.array(simfreqs), columns=[\"f_Y\", \"f_A\"])\n",
        "    inferred[\"nind\"] = n\n",
        "    simres = pd.concat([simres, inferred])\n",
        "\n",
        "\n",
        "  sns.set_style(\"whitegrid\")\n",
        "\n",
        "  sns.violinplot(simres, y=\"f_Y\", x=\"nind\", color=\"c\", cut=0)\n",
        "  plt.title(f\"Population frequencies, f_Y: {simfrq[0]}, f_A: {simfrq[1]}\")\n",
        "  plt.axhline(simfrq[0], color=\"y\", lw=1.2, zorder=3)\n",
        "  plt.ylim((-0.1, 1.1))\n",
        "  plt.show()\n",
        "\n",
        "  sns.violinplot(simres, y=\"f_A\", x=\"nind\", color=\"c\", cut=0)\n",
        "  plt.title(f\"Population frequencies, f_Y: {simfrq[0]}, f_A: {simfrq[1]}\")\n",
        "  plt.axhline(simfrq[1], color=\"y\", lw=1.2, zorder=3)\n",
        "  plt.ylim((-0.1, 1.1))\n",
        "  plt.show()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "IC9b9zNMntGg"
      },
      "outputs": [],
      "source": [
        "### QQplot test lrt\n",
        "### real data\n",
        "\n",
        "len_linked = len(res[res[\"pos\"] > 5.1e7])\n",
        "qunif = (np.arange(1, len_linked + 1) - 0.5) / len_linked # expected p-values\n",
        "\n",
        "#sns.scatterplot(x=np.sort(res[res[\"pos\"] > 5.1e7][\"p1\"]), y=qunif)\n",
        "#sns.lineplot(x=[0,1], y=[0,1])\n",
        "#plt.title(\"Real data\")\n",
        "#plt.xlabel(\"Observed p-values\")\n",
        "#plt.ylabel(\"Expected p-values\")\n",
        "#plt.show()\n",
        "\n",
        "sns.scatterplot(y=-np.log10(np.sort(res[res[\"pos\"] > 5.1e7][\"p1\"])), x=-np.log10(qunif))\n",
        "sns.lineplot(x=[0,4], y=[0,4])\n",
        "plt.title(\"Real data from unlinked region\")\n",
        "plt.xlabel(\"-log10(Expected p-values)\")\n",
        "plt.ylabel(\"-log10(Observed p-values)\")\n",
        "plt.show()\n",
        "\n",
        "### null models\n",
        "qunif = (np.arange(1, 10_001) - .5) / 10_000\n",
        "\n",
        "for sim_model in [0, 1, 2, 3]:\n",
        "  simgeno, _, simmale = sim(nind=25, model=sim_model)\n",
        "\n",
        "  for null_model in [1, 2, 3]:\n",
        "    p = []\n",
        "    for i in range(len(simgeno)):\n",
        "      llik = ll(converge(simgeno[i], simmale), simgeno[i], simmale)\n",
        "      if null_model == 1:\n",
        "        frq_null = np.mean(simgeno[i]) / 2\n",
        "        frq_null = np.array([frq_null, frq_null])\n",
        "      elif null_model == 2:\n",
        "        frq_null_0 = converge(simgeno[i], simmale, fixY=1e-4)\n",
        "        frq_null_1 = converge(simgeno[i], simmale, fixY=1-1e-4)\n",
        "        if ll(frq_null_0, simgeno[i], simmale) > ll(frq_null_1, simgeno[i], simmale):\n",
        "          frq_null = frq_null_0\n",
        "        else:\n",
        "          frq_null = frq_null_1\n",
        "\n",
        "      else:\n",
        "        frq_null_0 = converge(simgeno[i], simmale, fixA=1e-4)\n",
        "        frq_null_1 = converge(simgeno[i], simmale, fixA=1-1e-4)\n",
        "        if ll(frq_null_0, simgeno[i], simmale) > ll(frq_null_1, simgeno[i], simmale):\n",
        "          frq_null = frq_null_0\n",
        "        else:\n",
        "          frq_null = frq_null_1\n",
        "\n",
        "      llik_null = ll(frq_null, simgeno[i], simmale)\n",
        "      lrt = 2 * ( llik - llik_null)\n",
        "      p.append(1 - chi2.cdf(lrt, df=1))\n",
        "\n",
        "    sns.scatterplot(x=-np.log10(qunif), y=-np.log10(np.sort(p)))\n",
        "    sns.lineplot(x=[0,5], y=[0,5])\n",
        "    plt.title(f\"Null model {null_model} on data simulated under model {sim_model}\")\n",
        "    plt.xlabel(\"-log10(Expected p-values)\")\n",
        "    plt.ylabel(\"-log10(Observed p-values)\")\n",
        "    plt.show()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "3LrUmJztXUHg"
      },
      "outputs": [],
      "source": [
        "### test convergence\n",
        "randinit_frqs1 = []\n",
        "randinit_frqs2 = []\n",
        "for i in range(len(geno)):\n",
        "  randinit_frqs1.append(converge(geno[i], male, randomInit=True))\n",
        "  randinit_frqs2.append(converge(geno[i], male, randomInit=True))\n",
        "\n",
        "randinit_frqs1 = np.array(randinit_frqs1)\n",
        "randinit_frqs2 = np.array(randinit_frqs2)\n",
        "\n",
        "sns.scatterplot(x=randinit_frqs1[:,0], y=randinit_frqs2[:,0])\n",
        "sns.lineplot(x=[0,1], y=[0,1])\n",
        "plt.title(\"f_Y for each site with two different random initializations\")\n",
        "plt.xlabel(\"f_Y\")\n",
        "plt.ylabel(\"f_Y\")\n",
        "plt.show()\n",
        "\n",
        "sns.scatterplot(x=randinit_frqs1[:,1], y=randinit_frqs2[:,1])\n",
        "sns.lineplot(x=[0,1], y=[0,1])\n",
        "plt.title(\"f_A for each site with two different random initializations\")\n",
        "plt.xlabel(\"f_A\")\n",
        "plt.ylabel(\"f_A\")\n",
        "plt.show()\n",
        "\n",
        "diff = randinit_frqs1 - randinit_frqs2\n",
        "logdiff = np.log10(diff)[(diff[:, 0] > 1e-4) | (diff[:, 1] > 1e-4)]\n",
        "\n",
        "np.mean(diff[:20000, 0] > 1e-4)\n",
        "np.mean(diff[:20000, 1] > 1e-4)\n",
        "\n",
        "#sns.histplot(logdiff[:20000,0])\n",
        "#sns.histplot(logdiff[:20000,1])\n",
        "\n",
        "diff\n",
        "res[\"converged\"] = (diff[:, 0] < 5e-2) & (diff[:, 1] < 5e-2)\n",
        "np.mean(~res[\"converged\"])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "pU0a_JX6FyFV"
      },
      "outputs": [],
      "source": [
        "### playground block\n",
        "\n",
        "#np.unique(bed.compute(), return_counts=True)\n",
        "#np.unique(freq.compute(), return_counts=True)\n",
        "\n",
        "\n",
        "\n",
        "#EMInferFreqs(np.array([0.5, 0.25]), geno[0], sex, verbose=True)\n",
        "\n",
        "#converge(geno[4], sex)\n",
        "\n",
        "\n",
        "#print(np.array([geno[0], sex]).T)\n",
        "#multiprocessing.cpu_count()\n",
        "\n",
        "#res[res['f_Y']==0]\n",
        "#res[res[\"p1\"]>0.5]\n",
        "#converge(geno[6], male, fixA=True)\n",
        "#1 - chi2.cdf(40, df=1)\n",
        "res\n",
        "np.mean(res[\"sig2&3\"] & (res[\"pos\"] < 5.1e7))\n",
        "#int(np.rint(1 * 0.5))\n",
        "#np.array([True] * 5 + [False] * (10 - 5))\n",
        "\n",
        "sns.histplot(res[~res[\"converged\"]], x=\"pos\", bins=100)\n",
        "#sns.histplot(res, x=\"pos\", bins=100)\n",
        "plt.show()\n",
        "\n",
        "\n",
        "sns.jointplot(res[~res[\"converged\"]], x=\"f_Y\", y=\"f_A\", alpha = 0.05, marginal_ticks=True, ratio=2)\n",
        "plt.show()\n",
        "\n",
        "sns.jointplot(res[res[\"pos\"] < 5.1e7], x=\"f_Y\", y=\"f_A\", alpha = 0.05, marginal_ticks=True, ratio=2)\n",
        "plt.show()\n",
        "\n",
        "\n",
        "geno[res[\"sig2&3\"]][:, male][:5]\n",
        "geno[res[\"sig2&3\"]][:, ~male][:5]\n",
        "\n",
        "res"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyP9HD5orIANqLDthfI1hS9B",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}